USF-922: Extending drop-ins with Commerce APIs (#152)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/all/extending.mdx

import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';

Drop-in components are designed to be flexible and extensible. This guide provides an overview of how to extend drop-in components to add new features, integrate with third-party services, and customize the user experience.

## Extend drop-ins with Commerce APIs

> The following steps describe how to add existing Commerce API services to a drop-in. For example, the Commerce API provides the necessary endpoints to fetch and update gift messages through GraphQL, but the checkout drop-in doesn't provide this feature out of the box. We will extend the checkout drop-in by adding a UI for gift messages and then integrating the Commerce GraphQL API to fetch and update those messages.

Reviewer: ```suggestion
The following steps describe how to add existing Commerce API services to a drop-in. For example, the Commerce API provides the necessary endpoints to fetch and update gift messages through GraphQL, but the checkout drop-in doesn't provide this feature out of the box. We will extend the checkout drop-in by adding a UI for gift messages, use the Commerce GraphQL API to update the message data on the cart, and extend the cart drop-in to include the message data when it fetches the cart.
```

**Suggestion:** Added additional details to be more explicit on the upcoming steps.

### Step-by-step

<Tasks>

<Task>
### Add your UI to the drop-in

> The first step is to add a new UI component to the checkout drop-in. This component will allow users to enter a gift message for their order. We'll call this component `GiftOptionsField` and create it as a web component for fun as shown in the example implementation below.

Reviewer: **Question:** Do you think we should be more explicit that the UI can be implemented however the developer wants?

> The first step is to add a new UI component to the checkout drop-in. This component will allow users to enter a gift message for their order. We'll call this component `GiftOptionsField` and create it as a web component for fun as shown in the example implementation below.

Reviewer 2: ```suggestion
The first step is to add a new UI component to the checkout drop-in. This component will allow users to enter a gift message for their order. We'll call this component `GiftOptionsField` and create it as a web component, as shown in the example implementation below.
```

        UI.render(Input,
            {
                type: "text",
                name: "fromName",
                placeholder: "From name",
                floatingLabel: "From name",
                value: this.getAttribute('fromName'),
                disabled: !!this.hasAttribute('loading')
            })(fromNameWrapper);
        UI.render(TextArea,
            {
                name: "giftMessage",
                placeholder: "Message",
                value: this.getAttribute('giftMessage'),
                disabled: !!this.hasAttribute('loading')
            })(giftMessageWrapper);
        UI.render(Button,
            {
                variant: "primary",
                children: "Add Message",
                type: "submit",
                enabled: true,
                size: "medium",
                disabled: !!this.hasAttribute('loading')
            })(submitWrapper);

        this.shadowRoot.appendChild(sdkStyle.cloneNode(true));
        this.shadowRoot.appendChild(checkoutStyle.cloneNode(true));
    }
}

customElements.define('gift-options-field', GiftOptionsField);
```
</Task>

<Task>
### Render the UI into a checkout container

> Next, we need to render the `GiftOptionsField` component into the checkout drop-in by creating a container to render the component into.

Reviewer: ```suggestion
Next, we need to render the `GiftOptionsField` component into the checkout page by creating the `gift-options-field` [custom element](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements).
```

**Suggestion:** Update to be more specific.

```js
const GiftOptionsField = document.createElement('gift-options-field');
GiftOptionsField.setAttribute('loading', 'true');
```

> Then, add the container to the layouts of the checkout drop-in. The following example shows how to add the container to the `checkout` layout.

Reviewer: ```suggestion
Then, insert the custom element into the layouts defined on the checkout page. The following example updates the render function for mobile and desktop to insert the `giftOptionsField` element into the layouts.
```

**Suggestion:** Update to be more specific.

```js title='commerce-checkout.js'
function renderMobileLayout(block) {
  root.replaceChildren(
    heading,
    giftOptionsField,
    ...
  );

  block.replaceChildren(root);
}

function renderDesktopLayout(block) {
  main.replaceChildren(
    heading,
    giftOptionsField,
    ...
  );

  block.replaceChildren(block);
}
```

</Task>

<Task>
### Add handler for gift message submission

> Now that we have the UI in place, we need to add a handler to submit the gift message to the Commerce API. We'll use the `fetch` API to send a GraphQL mutation to update the gift message.

Reviewer: ```suggestion
Now that we have the UI in place, we need to add a handler to save the gift message data. We'll use the `fetchGraphl()` function from the API to send a GraphQL mutation to set the gift message on the cart.
```

**Suggestion:** Rewording to be more specific.

```js title='commerce-checkout.js'
giftOptionsField.submitGiftMessageHandler = async (event) => {
  event.preventDefault();

  const form = event.target;
  const formData = new FormData(form);
  const cartId = formData.get('cartId');
  const fromName = formData.get('fromName');
  const toName = formData.get('toName');
  const giftMessage = formData.get('giftMessage');

  giftOptionsField.setAttribute('loading', 'true');
  console.log('form data', cartId, fromName, toName, giftMessage);

  const giftMessageInput = {
    from: fromName,
    to: toName,
    message: giftMessage,
  }

  fetchGraphQl(`
    mutation SET_GIFT_OPTIONS($cartId: String!, $giftMessage: GiftMessageInput!) {
        setGiftOptionsOnCart(input: {
          cart_id: $cartId,
          gift_message: $giftMessage

>           gift_receipt_included: false

Reviewer 2: I wouldn't include `gift_receipt_included` or `printed_card_included` fields, as they're outside the scope of the example. If you mention them, then I feel like you need to expand the scope to include those fields in the UI.

          printed_card_included: false
        }) {
					cart {
						id
						gift_message {
							from
							to
							message
						}
					}
        }
      }
    `,
    {
    variables: {
      cartId,
      giftMessage: giftMessageInput,
    },
  }).then(() => {
    refreshCart();
    giftOptionsField.removeAttribute('loading');
  });
};
```
</Task>

<Task>
### Extend the data payload for the drop-in

> To extend the data payload of a drop-in, you need to add a GraphQL fragment to the `cart` payload by modifying the `build.mjs` script at the root of your storefront project. For our example, the fragment should add the gift message data in the cart. Here's an example of what the GraphQL should look like:

Reviewer: ```suggestion
To extend the data payload of a drop-in, first you need to update the GraphQL fragment used by the cart drop-in to request the additional field. This is done by modifying the `build.mjs` script at the root of your storefront project. In the following example, the `CART_FRAGMENT` fragment is extended to include the gift message data whenever the cart drop-in requests the cart data from GraphQL:
```

**Suggestion:** Update for clarification.

```js title='build.mjs'
/* eslint-disable import/no-extraneous-dependencies */
import { overrideGQLOperations } from '@dropins/build-tools/gql-extend.js';

// Extend the cart fragment to include the gift message
overrideGQLOperations([
  {
    // The name of the drop-in to extend
    npm: '@dropins/storefront-cart',
    // Additional fields to include in the cart results (gift_message)
    operations: [
      `fragment CART_FRAGMENT on Cart {
        gift_message {
          from
          to
          message
        }
      }`
    ],
  },
]);
```

> When you run the install command, the `build.mjs` script will generate a new GraphQL query for the cart drop-in that includes the `gift_message` data.  extend the cart fragment to include the gift message data.

Reviewer: **Suggestion:** We can probably remove that last sentence. It seems to be missing the beginning.

> When you run the install command, the `build.mjs` script will generate a new GraphQL query for the cart drop-in that includes the `gift_message` data.  extend the cart fragment to include the gift message data.

Reviewer 2: ```suggestion
When you run the install command, the `build.mjs` script generates a new GraphQL query for the cart drop-in that includes the `gift_message` data. 
```

</Task>

<Task>

### Add new data to the payload

> Map the updated GraphQL query to the data that the cart events provide to listeners. In this way, the checkout drop-in can access gift message data.

Reviewer: ```suggestion
Map the new GraphQL data to the payload data that the cart events provide to listeners so they can access the gift message values.
```

> Configure the cart drop-in's initializer to add the new cart data to the existing cart payload. This includes the new data with the payload that the cart events provide to listeners. As an example, here is how it might be configured:

Reviewer: ```suggestion
Configure the cart drop-in's initializer to add the new cart data to the existing cart payload. This is done by defining a transformer function on the CartModel. This function receives the GraphQL data and returns an object that gets merged with the rest of the cart payload. As an example, here is how it might be configured:
```

</Task>

<Task>

### Retrieve the data and render it

Get the data from the cart event and use it to populate the gift message fields on the checkout page. Here's an example of how you might do this:

```js title='commerce-checkout.js'
// Event listener to hydrate the new fields with the cart data
events.on('cart/data', data => {
  if (!data) return;

  const { id, orderAttributes, giftMessage } = data;

  // Update gift options fields
  giftOptionsField.setAttribute('cartId', id);
  if(giftMessage) {
    giftOptionsField.setAttribute('giftmessage', giftMessage.message);
    giftOptionsField.setAttribute('fromname', giftMessage.from);
    giftOptionsField.setAttribute('toname', giftMessage.to);
  }
  giftOptionsField.removeAttribute('loading');
}, { eager: true });
```

</Task>

<Task>

### Summary

After just a few changes, we were able to add a new feature to the checkout drop-in that allows users to add a gift message to their order. We added a new UI component, integrated the Commerce API to fetch and update gift messages, and extended the data payload for the drop-in to include the gift message data. You can apply these same concepts to any drop-in.

</Task>

</Tasks>

> ## Extend drop-ins with third-party components

Reviewer 2: Consider moving this section  to a separate topic.

Author: Yeah, I agree. I'll do that soon on another pass with a separate PR.
